### ISSUE TITLE
bind address causes connection refused on reboot

### ISSUE THREAD

**Issue by severach** [openSUSE/xinetd#13]
When a bind address is used, xinetd reports connection refused after a reboot. Xinetd connects properly when restarted after the address exists.

**Comment by vapier** [comment]
can you provide more details please ?  are you saying that if something has bound a specific port that xinetd is registered to use, it reports an error ?  or are you saying that if xinetd is configured to listen on a specific address, and that address isn't available when xinetd starts, it reports an error ?

**Comment by severach** [comment]
>xinetd is configured to listen on a specific address, and that address isn't available when xinetd starts, it reports an error 

That! Connections to the port show connection refused until I restart xinetd. This doesn't happen for services bound to all addresses.

**Comment by vapier** [comment]
tl;dr: what you describe is WAI today from the xinetd's project point of view.

i think currently that's expected.  xinetd doesn't poll for address changes on interfaces.  it's expected that the distro's init system would make sure the deps are configured such that xinetd is launched only after interfaces/addresses are available, and if the admin of the system adds more deps (like requiring additional addresses), they'd also adjust the init scripts to depend on whatever sets up those addresses.

so basically xinetd is dumb.  it starts up and tries to bind all the addresses/ports it's been configured for.  if an address/port combo are unavailable, it doesn't bind it.  if someone tries to connect to that address/port combo, then they'd get "connection refused" because xinetd isn't listening.

it's debatable whether xinetd should be more dynamic and listen to address changes (presumably via a netlink socket?) so it can automatically bind when things come available.  or perhaps add some internal polling logic so that if an address/port becomes available, it grabs it.  polling would be necessary for conflicts at the port level as there aren't (afaik) any event streams in Linux for when ports come & go.

being dynamic also complicates the security model.  we can't do all the privileged stuff up front (e.g. bind reserved ports) and then drop privs if we're going to be polling for it.  which means to maintain security, we'd have to do some privilege separation logic.
